### PR TITLE
Add improvements and fixes for benchmarks

### PR DESCRIPTION
**Pull request by plokhotnyuk**
- upgrade to latest versions of disruptor and JMH
- decrease number of iterations to speed up getting of result
- fix PI calculation in ForkJoin benchmarks
- add single thread benchmark
- minimize usage of float point operations in calculatePI function
- clean up of code

### CODE REVIEW

File: src/main/java/net/shipilev/Disruptor.java [hunk 1]
```diff
@@ -24,17 +24,17 @@
 import java.util.concurrent.TimeUnit;
 
 @State(Scope.Benchmark)
-@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
-@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
-@Fork(5)
+@Warmup(iterations = 3, time = 1, timeUnit = TimeUnit.MILLISECONDS)
```
Comment (shipilev) [inline src/main/java/net/shipilev/Disruptor.java]: Why 1 ms iterations?

File: src/main/java/net/shipilev/ForkJoin.java [hunk 2]
```diff
@@ -16,9 +16,9 @@
 import java.util.concurrent.TimeUnit;
 
 @State(Scope.Benchmark)
-@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
```
Comment (shipilev) [inline src/main/java/net/shipilev/ForkJoin.java]: Why 1 ms iterations?

File: src/main/java/net/shipilev/ForkJoin.java [hunk 3]
```diff
@@ -16,9 +16,9 @@
...
-@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
-@Fork(5)
+@Warmup(iterations = 3, time = 1, timeUnit = TimeUnit.MILLISECONDS)
```
Comment (shipilev) [inline src/main/java/net/shipilev/ForkJoin.java]: Why 1 ms iterations?

File: src/main/java/net/shipilev/ForkJoinRecursive.java [hunk 4]
```diff
@@ -15,9 +15,9 @@
 import java.util.concurrent.TimeUnit;
 
 @State(Scope.Benchmark)
-@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
-@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
-@Fork(5)
+@Warmup(iterations = 3, time = 1, timeUnit = TimeUnit.MILLISECONDS)
```
Comment (shipilev) [inline src/main/java/net/shipilev/ForkJoinRecursive.java]: Why 1 ms iterations?

File: src/main/java/net/shipilev/ForkJoinRecursiveDeep.java [hunk 5]
```diff
@@ -15,9 +15,9 @@
 import java.util.concurrent.TimeUnit;
 
 @State(Scope.Benchmark)
-@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
-@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
-@Fork(5)
+@Warmup(iterations = 3, time = 1, timeUnit = TimeUnit.MILLISECONDS)
```
Comment (shipilev) [inline src/main/java/net/shipilev/ForkJoinRecursiveDeep.java]: 1 ms iterations again.

File: src/main/java/net/shipilev/ForkJoinReuse.java [hunk 6]
```diff
@@ -2,25 +2,21 @@
...
-import java.util.List;
-import java.util.concurrent.ForkJoinPool;
 import java.util.concurrent.RecursiveTask;
 import java.util.concurrent.TimeUnit;
 
 @State(Scope.Benchmark)
-@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
-@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
-@Fork(5)
+@Warmup(iterations = 3, time = 1, timeUnit = TimeUnit.MILLISECONDS)
```
Comment (shipilev) [inline src/main/java/net/shipilev/ForkJoinReuse.java]: 1 ms iterations.

File: src/main/java/net/shipilev/Streams.java [hunk 7]
```diff
@@ -2,30 +2,28 @@
...
-import org.openjdk.jmh.annotations.State;
 import org.openjdk.jmh.annotations.Warmup;
 
 import java.util.concurrent.TimeUnit;
 import java.util.stream.IntStream;
 
-@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
-@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
-@Fork(5)
+@Warmup(iterations = 3, time = 1, timeUnit = TimeUnit.MILLISECONDS)
```
Comment (shipilev) [inline src/main/java/net/shipilev/Streams.java]: 1 ms iterations.

File: src/main/java/net/shipilev/ForkJoin.java (hunk 3 above)
Comment (plokhotnyuk) [inline src/main/java/net/shipilev/ForkJoin.java]: To set constant iteration numbers (3 for warming and 5 for result stat). 
If time is not set then JMH iterate several times until exceeds 1 second.

Comment (plokhotnyuk) [inline src/main/java/net/shipilev/ForkJoin.java]: Hi Alexey!

I have tried to print result of each benchmark before return with:
`System.out.println(acc + " ");`
And found that `@Warmup(iterations = 3)` and `@Measurement(iterations = 5)` produces more iteration than specified if their completion time less than 1 second.
Then after setting `time = 1, timeUnit = TimeUnit.MILLISECONDS` I got interesting results:

```
# Benchmark: net.shipilev.Disruptor.run

# Run progress: 0.00% complete, ETA 00:00:00
# Fork: 1 of 1
# Warmup Iteration   1: 384.449 ms/op
# Warmup Iteration   2: 435.573 ms/op
# Warmup Iteration   3: 433.340 ms/op
Iteration   1: 454.412 ms/op
Iteration   2: 440.380 ms/op
Iteration   3: 438.848 ms/op
Iteration   4: 436.011 ms/op
Iteration   5: 432.212 ms/op
...

# Benchmark: net.shipilev.SingleThread.run

# Run progress: 71.43% complete, ETA 00:00:07
# Fork: 1 of 1
# Warmup Iteration   1: 812.863 ms/op
# Warmup Iteration   2: 815.574 ms/op
# Warmup Iteration   3: 822.516 ms/op
Iteration   1: 815.691 ms/op
Iteration   2: 815.596 ms/op
Iteration   3: 815.986 ms/op
Iteration   4: 815.021 ms/op
Iteration   5: 829.840 ms/op

...

# Benchmark: net.shipilev.Streams.run

# Run progress: 85.71% complete, ETA 00:00:04
# Fork: 1 of 1
# Warmup Iteration   1: 555.185 ms/op
# Warmup Iteration   2: 340.118 ms/op
# Warmup Iteration   3: 338.866 ms/op
Iteration   1: 339.177 ms/op
Iteration   2: 341.390 ms/op
Iteration   3: 339.420 ms/op
Iteration   4: 339.130 ms/op
Iteration   5: 339.924 ms/op
```

For all benchmarks from this project 1 warmup iteration is enough and for some of them (Disruptor and SingleThread) best results are in 1st warmup iteration.

File: src/main/java/net/shipilev/Disruptor.java (hunk 1 above)
Comment (plokhotnyuk) [inline src/main/java/net/shipilev/Disruptor.java]: Just found that `@BenchmarkMode(Mode.SingleShotTime)` do what I tried to archive by `time = 1, timeUnit = TimeUnit.SECONDS` and with better reporting by percentile histogram.

Will you accept this PR with following annotations:

```
@State(Scope.Benchmark)
@Warmup(iterations = 3)
@Measurement(iterations = 10)
@Fork(1)
@BenchmarkMode(Mode.SingleShotTime)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
```

?

Main goal of switching from `AverageTime` to `SingleShotTime` mode is not hiding of per iteration results (by averaging). Minimal value does matter, especially when it occurs in first iteration.

**Comment by shipilev** [comment]
Yes, switching to `SingleShotTime` may be appropriate here, given how beefy the benchmarks are.

**Comment by shipilev** [comment]
Keep 3 warmup iterations though: smaller machines might require more time to warm up, and it's generally good to know the last warmup iteration has the same score as the first measurement one.

**Pull request merged**